File: api-xmlsignature/csharp/src/util/I18n.cs
using System;
using System.Reflection;
using System.Resources;
using System.Runtime.CompilerServices;
using System.Globalization;

namespace tr.gov.tubitak.uekae.esya.api.xmlsignature.util
{
    using Logger = log4net.ILog;
    using LogManager = log4net.LogManager;
	using I18nSettings = tr.gov.tubitak.uekae.esya.api.common.bundle.I18nSettings;

   	/// <summary>
	/// @author ahmety
	/// date: Jun 15, 2009
	/// </summary>
	public static class I18n
	{
        private static readonly ResourceManager mResources = new ResourceManager("tr.gov.tubitak.uekae.esya.api.xmlsignature.Properties.Resource", Assembly.GetExecutingAssembly()); 
		internal static readonly bool missingResources = false;
        internal static readonly Logger logger = LogManager.GetLogger(typeof(I18n));
	    internal static readonly CultureInfo mLocale;

		static I18n()
		{
			try
			{
			    mLocale = CultureInfo.CurrentCulture;
			}
			catch (Exception x)
			{
                logger.Warn("Error configuring locale from settings ", x);
			}
		}

          public static String message(String aMessage)
        {
            return (mResources.GetString(aMessage, I18nSettings.getLocale()));
        }
        public static String message(String aMessage, String[] aArgs)
        {
            //return message(String.Format(aMessage, aArgs));
            return String.Format(message(aMessage), aArgs);            
        }

		public static string translate(string aMessageId, params object[] aArgs)
		{
		    try
		    {
                if (aArgs == null || aArgs.Length == 0)
                {
                    return message(aMessageId);
                }
                else
                {
                    string[] args = new string[aArgs.Length];
                    for (int k = 0; k < aArgs.Length; k++)
                    {
                        args[k] = aArgs[k].ToString();
                    }
                    return message(aMessageId, (args));
                }

		    }
		    catch (Exception exc)
		    {
		        return aMessageId;
		    }
		    
		}
	}

}

File: api-xmlsignature/src/tr/gov/tubitak/uekae/esya/api/xmlsignature/util/I18n.java
package tr.gov.tubitak.uekae.esya.api.xmlsignature.util;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import tr.gov.tubitak.uekae.esya.api.common.ESYAException;
import tr.gov.tubitak.uekae.esya.api.common.bundle.I18nSettings;

import java.text.MessageFormat;
import java.util.Locale;
import java.util.ResourceBundle;

/**
 * @author ahmety
 * date: Jun 15, 2009
 */
public class I18n
{
    static ResourceBundle mBundle;
    static boolean missingResources = false;

    static Logger logger = LoggerFactory.getLogger(I18n.class);
    static Locale mLocale;

    static {
        try {
            mLocale = I18nSettings.getLocale();
        }
        catch (Exception x){
            logger.warn("Error configuring locale from settings ", x);
        }
        I18nSettings.addLocaleSetListener(new I18nSettings.LocaleChangeListener(){
            public void localeChanged(Locale aNewLocale)
            {
                mLocale = aNewLocale;
                init();
            }
        });
    }

    private static final String KAYNAK = "tr/gov/tubitak/uekae/esya/api/xmlsignature/resource/xmlimza";


    private static synchronized void init() {
        try {
            mBundle = ResourceBundle.getBundle(KAYNAK, mLocale);
        } catch (Exception x){
            missingResources = true;
            logger.error("Cant initialize locale successfully! ", x);
            mBundle = ResourceBundle.getBundle(KAYNAK);
        }
    }

    public static String translate(String aMessageId, Object... aArgs){
        if (mBundle==null && (!missingResources)){
            init();
        }
        try {
            if(mBundle == null)
             throw new ESYAException("mBundle should not be null");

            return MessageFormat.format(mBundle.getString(aMessageId), aArgs);

        } catch (Exception e){
            logger.warn("No message with ID '" + aMessageId + "' found in resource bundle '" + KAYNAK + "'", e);
            return aMessageId;
        }

    }

    public static void main(String[] args)
    {
        System.out.println(translate("unknown.algorithm","a"));
    }

}
